### ISSUE TITLE
PDF/A-2A validation CIDSet cidSetListsAllGlyphs issue for embedded font subset

### ISSUE THREAD

**Issue by evoronkin** [veraPDF/veraPDF-library#1258]
Pdf is produced with the iText framework with embedded Font with CIDSet.
VeraPDF validation fail evaluate the correct CIDSet for the last cid with report:

Specification: ISO 19005-2:2011, Clause: 6.2.11.4, Test number: 4   
If the FontDescriptor dictionary of an embedded CID font contains a CIDSet stream, then it shall identify all CIDs which are present in the font program, regardless of whether a CID in the font is referenced or used by the PDF or not. 

PDCIDFont   
fontFile_size == 0 || fontName.search(/[A-Z]{6}\+/) != 0 || CIDSet_size == 0 || cidSetListsAllGlyphs == true   
root/document[0]/pages[0](27 0 obj PDPage)/annots[0](53 0 obj PDWidgetAnnot)/appearance[0](69 0 obj PDContentStream)/operators[2]/xObject[0]/contentStream[0](68 0 obj PDContentStream)/operators[6]/xObject[0]/contentStream[0](66 0 obj PDContentStream)/operators[8]/font[0](IQJFBC+SourceSansPro-Regular)/DescendantFonts[0](IQJFBC+SourceSansPro-Regular)

I've debugged the issue in VeraPDF tool and found the cause in: 
https://github.com/veraPDF/veraPDF-validation/blob/integration/validation-model/src/main/java/org/verapdf/gf/model/impl/pd/font/GFPDCIDFont.java#168

Use of CID as Index is incorrect, it has to be corrected for the lookup. See all calls to bitSet.get(cid) in the method getcidSetListsAllGlyphs(). To make CID used as index it has to be (cid - 1)

Otherwise the lookup looks to the +1 index, so over the valid bitset place for the cid and 
then fails to match the last CID for the glyphs in the CIDSet.

For the reference how the iText creates the CIDSet, see:
https://github.com/itext/itext7-dotnet/blob/a3ce063817293592763c212248213bbdaa0fc53c/itext/itext.kernel/itext/kernel/font/PdfType0Font.cs#L737
Basically, one bit is set for each glyph.

To make it simple to explain:
Suppose we have only 1 glyph, so only one CID given. The CIDSet will contain only one 1 bit set with the index[0].
Now when the getcidSetListsAllGlyphs checks the bitSet.get(cid).
So for cid = 1, bitSet.get(cid) will return false, because it will lookup index[1], instead of index[0].

So based on, how iText creates the CIDSet, VeraPDF fail the above rule.
Please, consider to correct, not just one mentioned code place,
but all code where you call .get(cid) in the getcidSetListsAllGlyphs method.

**Comment by MaximPlusov** [comment]
Hi @evoronkin. Could you attach example of PDF file you describe in the issue?

**Comment by evoronkin** [comment]
[Demo filea.pdf](https://github.com/veraPDF/veraPDF-library/files/8815071/Demo.filea.pdf)

**Comment by bdoubrov** [comment]
@evoronkin thanks for the detailed analysis and the test file. The real issue here is in the embedded CFF font: it is an incorrect CharSet specifying the range of SIDs as 1-1942 (first SID=1 and 1941 left SIDs in terms of Table 20 of the CFF font specification: https://adobe-type-tools.github.io/font-tech-notes/pdfs/5176.CFF.pdf). 

So, veraPDF does assume that the font has 1942 glyphs in addition to the .notdef glyph with SID=0, which is always implicitly present. This does not match the CIDSet in the font dictionary, which results in the validation error.

In fact, other data in the font such as Charstrings index count show that the font contains only 1941 glyph. As a workaround, we'll probably add this to the veraPDF logic to trim the CharSet range by the number of glyphs specified in the other font data. I guess this is what Adobe implementation is doing now, as it does not report this validation error.

**Comment by MatthiasValvekens** [comment]
Yes, this is a bug in iText, not VeraPDF. I ran across it a couple of weeks ago and reported it internally. The bug's pretty impressively old, though: it dates back to when CFF subsetting support was first added to iText (almost two decades ago!).

**Comment by evoronkin** [comment]
@MatthiasValvekens could you give a reference to the fix in the iText which is pending, thx.

**Comment by MatthiasValvekens** [comment]
It's on our internal bug tracker, which unfortunately isn't public. If you want to be notified when the fix is merged, please file a support ticket and request that it be linked to the corresponding dev ticket. If you link this thread, our support staff will know what to do.

**Comment by PaulLoth** [comment]
Linked this to itext support ticket SUP-35940 for reference.

**Comment by evoronkin** [comment]
Great job @MaximPlusov your fix in the latest VeraPdf works well!
